File: client-libraries/python/audit_trail_client.py
"""
Audit Trail Client Library for Python

Library untuk integrasi dengan Audit Trail API
Bisa digunakan di Django, Flask, FastAPI, dll

@version 1.0.0
@author Audit Trail System
"""

import requests
import time
from typing import Optional, Dict, Any


class AuditTrailClient:
    """
    Client untuk komunikasi dengan Audit Trail API
    """
    
    def __init__(self, base_url: str, user_id: str, password: str, debug: bool = False):
        """
        Constructor
        
        Args:
            base_url: URL base audit trail (contoh: http://localhost/audit-trail/api)
            user_id: ID aplikasi dari registrasi
            password: Password aplikasi
            debug: Enable debug mode (default: False)
        """
        self.base_url = base_url.rstrip('/')
        self.user_id = user_id
        self.password = password
        self.token = None
        self.token_expiry = None
        self.debug = debug
    
    def _get_token(self) -> Optional[str]:
        """
        Get atau refresh token
        
        Returns:
            Token atau None jika gagal
        """
        # Jika token masih valid, return token yang ada
        if self.token and self.token_expiry and time.time() < self.token_expiry:
            return self.token
        
        try:
            response = requests.get(
                f'{self.base_url}/getToken',
                headers={
                    'x-userid': self.user_id,
                    'x-password': self.password
                },
                timeout=10
            )
            
            data = response.json()
            
            if self.debug:
                print(f'Audit Trail - Get Token Response: {data}')
            
            if response.status_code == 200 and 'response' in data and 'token' in data['response']:
                self.token = data['response']['token']
                # Token berlaku 15 menit, kita set 14 menit untuk safety
                self.token_expiry = time.time() + 840
                return self.token
            
            print(f'Audit Trail - Failed to get token: {data}')
            return None
            
        except Exception as e:
            print(f'Audit Trail - Error getting token: {e}')
            return None
    
    def log(self, user: str, aksi: str, **options) -> Dict[str, Any]:
        """
        Kirim log aktivitas ke audit trail
        
        Args:
            user: Username yang melakukan aktivitas
            aksi: Jenis aksi (CREATE, UPDATE, DELETE, LOGIN, LOGOUT, dll)
            **options: Optional parameters:
                - menu_fitur: Nama menu/fitur (str)
                - hasil: success atau failed (default: success)
                - no_rm: Nomor rekam medis (str)
                - rawat: JALAN atau INAP (str)
                - trx_id: ID transaksi (str)
                - ip_address: IP address user (str, auto-detect jika kosong)
                - ket: Keterangan custom (str, auto-generate jika kosong)
        
        Returns:
            Dict dengan keys: success, message, code, data
        """
        token = self._get_token()
        
        if not token:
            return {
                'success': False,
                'message': 'Failed to get token',
                'code': 0
            }
        
        # Build data
        data = {
            'user': user,
            'aksi': aksi.upper()
        }
        
        # Merge optional parameters
        if 'menu_fitur' in options:
            data['menu_fitur'] = options['menu_fitur']
        if 'hasil' in options:
            data['hasil'] = options['hasil']
        if 'no_rm' in options:
            data['no_rm'] = options['no_rm']
        if 'rawat' in options:
            data['rawat'] = options['rawat'].upper()
        if 'trx_id' in options:
            data['trx_id'] = options['trx_id']
        if 'ip_address' in options:
            data['ip_address'] = options['ip_address']
        if 'ket' in options:
            data['ket'] = options['ket']
        
        try:
            response = requests.post(
                f'{self.base_url}/postToken',
                headers={
                    'x-token': token,
                    'Content-Type': 'application/json'
                },
                json={'data': data},
                timeout=10
            )
            
            result = response.json()
            
            if self.debug:
                print(f'Audit Trail - Log Response: {result}')
            
            # Jika token expired, reset dan retry sekali
            if response.status_code == 401 and 'metadata' in result:
                if 'expired' in result['metadata'].get('message', '').lower():
                    self.token = None
                    self.token_expiry = None
                    
                    # Retry sekali
                    return self.log(user, aksi, **options)
            
            return {
                'success': response.status_code == 200,
                'message': result.get('metadata', {}).get('message', 'Unknown error'),
                'code': response.status_code,
                'data': result
            }
            
        except Exception as e:
            print(f'Audit Trail - Error sending log: {e}')
            return {
                'success': False,
                'message': str(e),
                'code': 0
            }
    
    def log_create(self, user: str, menu: str, keterangan: Optional[str] = None) -> Dict[str, Any]:
        """Shortcut untuk log CREATE"""
        return self.log(user, 'CREATE', menu_fitur=menu, ket=keterangan)
    
    def log_update(self, user: str, menu: str, keterangan: Optional[str] = None) -> Dict[str, Any]:
        """Shortcut untuk log UPDATE"""
        return self.log(user, 'UPDATE', menu_fitur=menu, ket=keterangan)
    
    def log_delete(self, user: str, menu: str, keterangan: Optional[str] = None) -> Dict[str, Any]:
        """Shortcut untuk log DELETE"""
        return self.log(user, 'DELETE', menu_fitur=menu, ket=keterangan)
    
    def log_login(self, user: str) -> Dict[str, Any]:
        """Shortcut untuk log LOGIN"""
        return self.log(
            user, 
            'LOGIN', 
            menu_fitur='Authentication',
            ket=f'{user} berhasil login ke sistem'
        )
    
    def log_logout(self, user: str) -> Dict[str, Any]:
        """Shortcut untuk log LOGOUT"""
        return self.log(
            user,
            'LOGOUT',
            menu_fitur='Authentication',
            ket=f'{user} telah logout dari sistem'
        )
    
    def enable_debug(self):
        """Enable debug mode"""
        self.debug = True
    
    def disable_debug(self):
        """Disable debug mode"""
        self.debug = False

File: client-libraries/javascript/AuditTrailClient.js
/**
 * Audit Trail Client Library for JavaScript
 * 
 * Library untuk integrasi dengan Audit Trail API
 * Bisa digunakan di Node.js atau Browser
 * 
 * @version 1.0.0
 * @author Audit Trail System
 */

class AuditTrailClient {
    
    /**
     * Constructor
     * 
     * @param {string} baseUrl - URL base audit trail (contoh: http://localhost/audit-trail/api)
     * @param {string} userId - ID aplikasi dari registrasi
     * @param {string} password - Password aplikasi
     * @param {boolean} debug - Enable debug mode (default: false)
     */
    constructor(baseUrl, userId, password, debug = false) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.userId = userId;
        this.password = password;
        this.token = null;
        this.tokenExpiry = null;
        this.debug = debug;
    }
    
    /**
     * Get atau refresh token
     * 
     * @returns {Promise<string|null>} Token atau null jika gagal
     */
    async getToken() {
        // Jika token masih valid, return token yang ada
        if (this.token && this.tokenExpiry && Date.now() < this.tokenExpiry) {
            return this.token;
        }
        
        try {
            const response = await fetch(`${this.baseUrl}/getToken`, {
                method: 'GET',
                headers: {
                    'x-userid': this.userId,
                    'x-password': this.password
                }
            });
            
            const data = await response.json();
            
            if (this.debug) {
                console.log('Audit Trail - Get Token Response:', data);
            }
            
            if (response.ok && data.response?.token) {
                this.token = data.response.token;
                // Token berlaku 15 menit, kita set 14 menit untuk safety
                this.tokenExpiry = Date.now() + (840 * 1000);
                return this.token;
            }
            
            console.error('Audit Trail - Failed to get token:', data);
            return null;
            
        } catch (error) {
            console.error('Audit Trail - Error getting token:', error);
            return null;
        }
    }
    
    /**
     * Kirim log aktivitas ke audit trail
     * 
     * @param {string} user - Username yang melakukan aktivitas
     * @param {string} aksi - Jenis aksi (CREATE, UPDATE, DELETE, LOGIN, LOGOUT, dll)
     * @param {Object} options - Optional parameters:
     *   - menu_fitur: Nama menu/fitur (string)
     *   - hasil: success atau failed (default: success)
     *   - no_rm: Nomor rekam medis (string)
     *   - rawat: JALAN atau INAP (string)
     *   - trx_id: ID transaksi (string)
     *   - ip_address: IP address user (string, auto-detect jika kosong)
     *   - ket: Keterangan custom (string, auto-generate jika kosong)
     * 
     * @returns {Promise<Object>} Response dari API
     */
    async log(user, aksi, options = {}) {
        const token = await this.getToken();
        
        if (!token) {
            return {
                success: false,
                message: 'Failed to get token',
                code: 0
            };
        }
        
        // Build data
        const data = {
            user: user,
            aksi: aksi.toUpperCase(),
            ...options
        };
        
        // Uppercase rawat jika ada
        if (data.rawat) {
            data.rawat = data.rawat.toUpperCase();
        }
        
        try {
            const response = await fetch(`${this.baseUrl}/postToken`, {
                method: 'POST',
                headers: {
                    'x-token': token,
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ data })
            });
            
            const result = await response.json();
            
            if (this.debug) {
                console.log('Audit Trail - Log Response:', result);
            }
            
            // Jika token expired, reset dan retry sekali
            if (response.status === 401 && result.metadata?.message?.toLowerCase().includes('expired')) {
                this.token = null;
                this.tokenExpiry = null;
                
                // Retry sekali
                return await this.log(user, aksi, options);
            }
            
            return {
                success: response.ok,
                message: result.metadata?.message || 'Unknown error',
                code: response.status,
                data: result
            };
            
        } catch (error) {
            console.error('Audit Trail - Error sending log:', error);
            return {
                success: false,
                message: error.message,
                code: 0
            };
        }
    }
    
    /**
     * Shortcut untuk log CREATE
     */
    async logCreate(user, menu, keterangan = null) {
        return await this.log(user, 'CREATE', {
            menu_fitur: menu,
            ket: keterangan
        });
    }
    
    /**
     * Shortcut untuk log UPDATE
     */
    async logUpdate(user, menu, keterangan = null) {
        return await this.log(user, 'UPDATE', {
            menu_fitur: menu,
            ket: keterangan
        });
    }
    
    /**
     * Shortcut untuk log DELETE
     */
    async logDelete(user, menu, keterangan = null) {
        return await this.log(user, 'DELETE', {
            menu_fitur: menu,
            ket: keterangan
        });
    }
    
    /**
     * Shortcut untuk log LOGIN
     */
    async logLogin(user) {
        return await this.log(user, 'LOGIN', {
            menu_fitur: 'Authentication',
            ket: `${user} berhasil login ke sistem`
        });
    }
    
    /**
     * Shortcut untuk log LOGOUT
     */
    async logLogout(user) {
        return await this.log(user, 'LOGOUT', {
            menu_fitur: 'Authentication',
            ket: `${user} telah logout dari sistem`
        });
    }
    
    /**
     * Enable debug mode
     */
    enableDebug() {
        this.debug = true;
    }
    
    /**
     * Disable debug mode
     */
    disableDebug() {
        this.debug = false;
    }
}

// Export untuk Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = AuditTrailClient;
}
